### DISCUSSION TITLE
Find first and last visible Item in a List using virtualization

### DISCUSSION THREAD

**Discussion by agnauck** [AvaloniaUI/Avalonia#10445] [General]
I need to find the first and last visible item in a long vertical scrollable ListBox which is using a VirtualizingStackPanel for better performance.

This is what I am trying to do currently:
* I am using the Childen collection of the VirtualizingStackPanel
* The ordering of the childs does not match the order of the UI, so I am sorting them by position (Bounds.Top)
* Then I am filtering the Children by the IsVisible property to get only those which should be visible in the virtual panel currently
* Once I have the list in order filtered by IsVisible I can just get the First and Last item

Te above approach seems to work fine when I scroll down the list. When I scroll up again and items disappear it looks like the IsVisible property is not getting reverted to false for some of them that disappear in the UI.

I am wondering if this is a bug or my approach is wrong and there is a better solution to what I am trying to achieve. But I have not found a simple Member like FirstVisibleItem or LastVisibleItem.

My Items have variable heights.

Thanks,
Alex

**Comment by maxkatz6** [comment]
> When I scroll up again and items disappear it looks like the IsVisible property is not getting reverted to false for some of them that disappear in the UI.

It won't, because this property defines if control should be rendered or not. It doesn't return if control is actually visible on the screen right now or not.

Better approach would be to go through sorted Childen collection and check which of them are inside of the scroll viewer viewport.
I haven't done it in a while, but I think you should be able to get scroll viewer and read its properties using `var scrollViewer = listBox.GetVisualDescendants().OfType<ScrollViewer>();` snipped (note, listbox should have applied template for that).

**Comment by grokys** [comment]
Doesn't `ItemsControl.GetRealizedContainers` work for your use-case? (Assuming that you're using a recent 11.0 preview)

**Reply by agnauck** [reply]
Thanks @maxkatz6 @grokys 
Both your suggestions work fine. I think I will go with the `ItemsControl.GetRealizedContainers()`

Alex
